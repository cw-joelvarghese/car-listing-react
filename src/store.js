import { configureStore } from "@reduxjs/toolkit";
import carsReducers from "./reducers/carsReducer";
import filterReducer from "./reducers/filterReducer";

export default configureStore({
    reducer: {
        cars: carsReducers,
        filter: filterReducer
    },
});
