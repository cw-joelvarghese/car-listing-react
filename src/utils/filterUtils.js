export const FUEL_TYPES = {
    PETROL: 1,
    DIESEL: 2,
    CNG: 3,
    LPG: 4,
    ELECTRIC: 5,
    HYBRID: 6,
};

export const SORT_TYPES = {
    UNSORTED: 1,
    PRICE_LOW_TO_HIGH: 2,
    PRICE_HIGH_TO_LOW: 3,
};

export const defaultFilterObject = {
    fuel: [],
    budgetStart: null,
    budgetEnd: null,
    sort: SORT_TYPES.UNSORTED,
};

export function sortCars(cars, filter) {
    if (filter.sort === SORT_TYPES.UNSORTED) {
        return cars;
    }
    if (filter.sort === SORT_TYPES.PRICE_HIGH_TO_LOW) {
        return cars.sort(function (a, b) {
            var priceA = a.priceNumeric,
                priceB = b.priceNumeric;
            // Compare the 2 dates
            return priceB - priceA;
        });
    }
    if (filter.sort === SORT_TYPES.PRICE_LOW_TO_HIGH) {
        return cars.sort(function (a, b) {
            var priceA = a.priceNumeric,
                priceB = b.priceNumeric;
            // Compare the 2 dates
            return priceA - priceB;
        });
    }
}

export function createFilterString(filter) {
    const searchFilter = {};
    if (filter.fuel.length > 0) {
        searchFilter.fuel = filter.fuel.join("+");
    }
    if (
        typeof filter.budgetStart == "number" &&
        typeof filter.budgetEnd == "number"
    ) {
        searchFilter.budget = `${filter.budgetStart}-${filter.budgetEnd}`;
    }
    return decodeURIComponent(new URLSearchParams(searchFilter).toString()).toString()
}