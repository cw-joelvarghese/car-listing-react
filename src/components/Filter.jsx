import { CarContext, FUEL_TYPES } from "src/App";
import styles from "./Filter.module.css";
import { useContext } from "react";

function removeItemFromArray(array, item) {
    var index = array.indexOf(item);
    if (index !== -1) {
        array.splice(index, 1);
    }
    return array;
}

function FuelCheckboxFilter({ fuelType, fuelName }) {
    const { filter, setFilter } = useContext(CarContext);
    function onClick(e) {
        const checked = e.target.checked;
        if (checked) {
            setFilter((prev) => ({
                ...prev,
                fuel: [...prev.fuel, fuelType],
            }));
            return;
        }
        setFilter((prev) => ({
            ...prev,
            fuel: removeItemFromArray([...prev.fuel], fuelType),
        }));
    }
    return (
        <CheckboxFilterItem
            name={fuelName}
            onClick={onClick}
            checked={filter.fuel.includes(fuelType)}
        />
    );
}

function CheckboxFilterItem({ name, onClick, checked }) {
    let r = (Math.random() + 1).toString(36).substring(7);
    return (
        <div className={styles.checkboxContainer}>
            <input
                type="checkbox"
                checked={checked}
                id={r}
                className={styles.checkbox}
                onChange={onClick}
            />
            <label className={styles.checkboxContainer} htmlFor={r}>
                {name}
            </label>
        </div>
    );
}

function BudgetFilter() {
    const { filter, setFilter } = useContext(CarContext);

    function onLowerChange(e) {
        setFilter((prev) => ({
            ...prev,
            budgetStart: Number(e.target.value)
        }))
    }

    function onUpperChange(e) {
        setFilter((prev) => ({
            ...prev,
            budgetEnd: Number(e.target.value)
        }))
    }

    return (
        <div className={styles.budgetContainer}>
            <input type="number" onChange={onLowerChange}/>
            <p>-</p>
            <input type="number" onChange={onUpperChange}/>
        </div>
    );
}

function Filter() {
    return (
        <div className={styles.container}>
            <div className={styles.headingContainer}>
                <div>Filters</div>
                <div>Clear All</div>
            </div>
            <div className={styles.checkboxesContainer}>
                <p>Fuel</p>
                {Object.keys(FUEL_TYPES).map((fuelType) => (
                    <FuelCheckboxFilter
                        key={fuelType}
                        fuelName={fuelType}
                        fuelType={FUEL_TYPES[fuelType]}
                    />
                ))}
                <p>Budget</p>
                <BudgetFilter />
            </div>
        </div>
    );
}

export default Filter;
