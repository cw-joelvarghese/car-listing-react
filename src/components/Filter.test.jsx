import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import Filter, {
    BudgetFilter,
    CheckboxFilterItem,
    FuelCheckboxFilter,
} from "src/components/Filter";
import store from "src/store";
import { FUEL_TYPES } from "src/utils/appUtils";
import { expect, test } from "vitest";
import '@testing-library/jest-dom'


test("Filter component renders correctly", () => {
    const tree = render(
        <Provider store={store}>
            <Filter />
        </Provider>
    );
    expect(tree).toMatchSnapshot();
});

test("BudgetFilter component renders correctly", () => {
    const tree = render(
        <Provider store={store}>
            <BudgetFilter />
        </Provider>
    );
    expect(tree).toMatchSnapshot();
});

test("CheckboxFilterItem component renders correctly", () => {
    const tree = render(
        <CheckboxFilterItem
            name={"PETROL"}
            onClick={() => {}}
            checked={false}
        />
    );
    expect(tree).toMatchSnapshot();
});

test("FuelCheckboxFilter component renders correctly", () => {
    const tree = render(
        <Provider store={store}>
            <FuelCheckboxFilter fuelType={FUEL_TYPES.PETROL} fuelName={"PETROL"} />
        </Provider>
    );
    expect(tree.getByText(/Petrol/i)).toBeInTheDocument()
    expect(tree).toMatchSnapshot();
});
