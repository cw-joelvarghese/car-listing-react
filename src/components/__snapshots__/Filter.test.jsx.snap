// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BudgetFilter component renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_budgetContainer_e8fa16"
      >
        <input
          type="number"
          value=""
        />
        <p>
          -
        </p>
        <input
          type="number"
          value=""
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_budgetContainer_e8fa16"
    >
      <input
        type="number"
        value=""
      />
      <p>
        -
      </p>
      <input
        type="number"
        value=""
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CheckboxFilterItem component renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_checkboxContainer_e8fa16"
      >
        <input
          class="_checkbox_e8fa16"
          id="PETROL"
          type="checkbox"
        />
        <label
          class="_checkboxLabel_e8fa16"
          for="PETROL"
        >
          PETROL
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_checkboxContainer_e8fa16"
    >
      <input
        class="_checkbox_e8fa16"
        id="PETROL"
        type="checkbox"
      />
      <label
        class="_checkboxLabel_e8fa16"
        for="PETROL"
      >
        PETROL
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Filter component renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_e8fa16"
      >
        <div
          class="_headingContainer_e8fa16"
        >
          <div>
            Filters
          </div>
          <button
            class="_clearButton_e8fa16"
          >
            Clear All
          </button>
        </div>
        <div
          class="_checkboxesContainer_e8fa16"
        >
          <p>
            Fuel
          </p>
          <div
            class="_checkboxContainer_e8fa16"
          >
            <input
              class="_checkbox_e8fa16"
              id="PETROL"
              type="checkbox"
            />
            <label
              class="_checkboxLabel_e8fa16"
              for="PETROL"
            >
              PETROL
            </label>
          </div>
          <div
            class="_checkboxContainer_e8fa16"
          >
            <input
              class="_checkbox_e8fa16"
              id="DIESEL"
              type="checkbox"
            />
            <label
              class="_checkboxLabel_e8fa16"
              for="DIESEL"
            >
              DIESEL
            </label>
          </div>
          <div
            class="_checkboxContainer_e8fa16"
          >
            <input
              class="_checkbox_e8fa16"
              id="CNG"
              type="checkbox"
            />
            <label
              class="_checkboxLabel_e8fa16"
              for="CNG"
            >
              CNG
            </label>
          </div>
          <div
            class="_checkboxContainer_e8fa16"
          >
            <input
              class="_checkbox_e8fa16"
              id="LPG"
              type="checkbox"
            />
            <label
              class="_checkboxLabel_e8fa16"
              for="LPG"
            >
              LPG
            </label>
          </div>
          <div
            class="_checkboxContainer_e8fa16"
          >
            <input
              class="_checkbox_e8fa16"
              id="ELECTRIC"
              type="checkbox"
            />
            <label
              class="_checkboxLabel_e8fa16"
              for="ELECTRIC"
            >
              ELECTRIC
            </label>
          </div>
          <div
            class="_checkboxContainer_e8fa16"
          >
            <input
              class="_checkbox_e8fa16"
              id="HYBRID"
              type="checkbox"
            />
            <label
              class="_checkboxLabel_e8fa16"
              for="HYBRID"
            >
              HYBRID
            </label>
          </div>
          <p>
            Budget
          </p>
          <div
            class="_budgetContainer_e8fa16"
          >
            <input
              type="number"
              value=""
            />
            <p>
              -
            </p>
            <input
              type="number"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_e8fa16"
    >
      <div
        class="_headingContainer_e8fa16"
      >
        <div>
          Filters
        </div>
        <button
          class="_clearButton_e8fa16"
        >
          Clear All
        </button>
      </div>
      <div
        class="_checkboxesContainer_e8fa16"
      >
        <p>
          Fuel
        </p>
        <div
          class="_checkboxContainer_e8fa16"
        >
          <input
            class="_checkbox_e8fa16"
            id="PETROL"
            type="checkbox"
          />
          <label
            class="_checkboxLabel_e8fa16"
            for="PETROL"
          >
            PETROL
          </label>
        </div>
        <div
          class="_checkboxContainer_e8fa16"
        >
          <input
            class="_checkbox_e8fa16"
            id="DIESEL"
            type="checkbox"
          />
          <label
            class="_checkboxLabel_e8fa16"
            for="DIESEL"
          >
            DIESEL
          </label>
        </div>
        <div
          class="_checkboxContainer_e8fa16"
        >
          <input
            class="_checkbox_e8fa16"
            id="CNG"
            type="checkbox"
          />
          <label
            class="_checkboxLabel_e8fa16"
            for="CNG"
          >
            CNG
          </label>
        </div>
        <div
          class="_checkboxContainer_e8fa16"
        >
          <input
            class="_checkbox_e8fa16"
            id="LPG"
            type="checkbox"
          />
          <label
            class="_checkboxLabel_e8fa16"
            for="LPG"
          >
            LPG
          </label>
        </div>
        <div
          class="_checkboxContainer_e8fa16"
        >
          <input
            class="_checkbox_e8fa16"
            id="ELECTRIC"
            type="checkbox"
          />
          <label
            class="_checkboxLabel_e8fa16"
            for="ELECTRIC"
          >
            ELECTRIC
          </label>
        </div>
        <div
          class="_checkboxContainer_e8fa16"
        >
          <input
            class="_checkbox_e8fa16"
            id="HYBRID"
            type="checkbox"
          />
          <label
            class="_checkboxLabel_e8fa16"
            for="HYBRID"
          >
            HYBRID
          </label>
        </div>
        <p>
          Budget
        </p>
        <div
          class="_budgetContainer_e8fa16"
        >
          <input
            type="number"
            value=""
          />
          <p>
            -
          </p>
          <input
            type="number"
            value=""
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`FuelCheckboxFilter component renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_checkboxContainer_e8fa16"
      >
        <input
          class="_checkbox_e8fa16"
          id="PETROL"
          type="checkbox"
        />
        <label
          class="_checkboxLabel_e8fa16"
          for="PETROL"
        >
          PETROL
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_checkboxContainer_e8fa16"
    >
      <input
        class="_checkbox_e8fa16"
        id="PETROL"
        type="checkbox"
      />
      <label
        class="_checkboxLabel_e8fa16"
        for="PETROL"
      >
        PETROL
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
